### ISSUE TITLE
[Bug]: Generates wrong JS for StringLiteral reused from JSXAttribute.value

### ISSUE THREAD

**Issue by andersk** [babel/babel#15472]
### 💻

- [ ] Would you like to work on a fix?

### How are you using Babel?

Programmatic API (`babel.transform`, `babel.parse`)

### Input code

This test case extracts the `StringLiteral` node from the `value` of a `JSXAttribute` node, and reuses it to generate some JavaScript code:

```js
const { default: generate } = require("@babel/generator");
const { parse } = require("@babel/parser");
const t = require("@babel/types");

const code = '<b title="&amp; \\xxx" />';
const ast = parse(code, { plugins: ["jsx"] });
const value = ast.program.body[0].expression.openingElement.attributes[0].value;
console.log(generate(t.callExpression(t.identifier("setTitle"), [value])).code);
```

### Current and expected behavior

Babel reuses `value.extra.rawValue`, which was encoded for HTML and not for JavaScript, and outputs this invalid JavaScript code:

```js
setTitle("&amp; \xxx")
```

Expected:

```js
setTitle("& \\xxx")
```

(or perhaps an error from `@babel/generator`).

## System:
 - OS: Linux 6.2 NixOS 23.05 (Stoat) 23.05 (Stoat)
## Binaries:
 - Node: 18.14.2 - ~/.nix-profile/bin/node
 - Yarn: 1.22.19 - ~/.nix-profile/bin/yarn
 - npm: 9.5.0 - ~/.nix-profile/bin/npm
## npmPackages:
 - @babel/generator: ^7.21.1 => 7.21.1 
 - @babel/parser: ^7.21.2 => 7.21.2 
 - @babel/types: ^7.21.2 => 7.21.2 

### Possible solution

I can think of three solutions.

1. Change the type of `JSXAttribute.value` literals to a new type, say `JSXStringLiteral`, that can’t be confused with `StringLiteral`. This would be a backwards-incompatible change.

2. When parsing a JSX attribute literal, mark the `StringLiteral` with a new attribute, say `.extra.jsxAttribute = true`. When generating code for a `StringLiteral` as a JSX attribute, only reuse `.extra.rawValue` when `.extra.jsxAttribute` is `true`; when generating code for a `StringLiteral` anywhere else, only reuse `.extra.rawValue` when `.extra.jsxAttribute` is not `true`.

3. When generating code for a `StringLiteral`, re-parse `.extra.rawValue` for the present context and check that it actually matches `.value` before reusing it.

### Additional context

- https://github.com/solidjs/solid/issues/1553

**Comment by andersk** [comment]
The converse bug where we generate invalid JSX from a `StringLiteral` that didn’t come from JSX can be reproduced even when the `StringLiteral` is fresh and not reused:

```js
const { default: generate } = require("@babel/generator");
const t = require("@babel/types");

console.log(
  generate(
    t.jsxElement(
      t.jsxOpeningElement(t.jsxIdentifier("b"), [
        t.jsxAttribute(t.jsxIdentifier("title"), t.stringLiteral('"')),
      ]),
      t.jsxClosingElement(t.jsxIdentifier("b")),
      []
    )
  ).code
);
```

Outputs this invalid JSX:

```jsx
<b title="\""></b>
```

Expected:

```jsx
<b title="&quot;"></b>
```

I think this argues in favor of solution 1: a JSX attribute literal should be a different node type from `StringLiteral` because it needs to be serialized differently.

**Comment by liuxingbaoyu** [comment]
Yes, it looks like solution 1 is the best, but I'm afraid there will be some trouble with compatibility.
Solution 3 looks good, we can store only `jsxRaw` in `.extra` for jsx nodes to reduce performance impact.

Let's see what other members think.
